projects/RabbitMQ.Client/client/impl/OutgoingCommand.cs

```diff
@@ -31,49 +31,74 @@
+    }
 
-        public readonly MethodBase Method;
-        private readonly ContentHeaderBase _header;
-        private readonly ReadOnlyMemory<byte> _body;
+    internal readonly struct OutgoingCommand : IOutgoingCommand
+    {
+        private readonly MethodBase _method;
+        public MethodBase Method => _method;
 
         public OutgoingCommand(MethodBase method)
-            : this(method, null, ReadOnlyMemory<byte>.Empty)
         {
+            _method = method;
         }
 
-        public OutgoingCommand(MethodBase method, ContentHeaderBase header, ReadOnlyMemory<byte> body)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly ReadOnlyMemory<byte> SerializeToFrames(ushort channelNumber, uint frameMax)
+        {
+            int size = Method.GetRequiredBufferSize() + Framing.Method.FrameSize;
+
+            // Will be returned by SocketFrameWriter.WriteLoop
+            byte[] rentedArray = ArrayPool<byte>.Shared.Rent(size);
+            int offset = Framing.Method.WriteTo(rentedArray, channelNumber, Method);
+#if DEBUG
+            System.Diagnostics.Debug.Assert(offset == size, $"Serialized to wrong size, expect {size}, offset {offset}");
+#endif
+
+            return new ReadOnlyMemory<byte>(rentedArray, 0, size);
+        }
+    }
+
+    internal readonly struct OutgoingContentCommand : IOutgoingCommand
+    {
+        public MethodBase Method => _method;
+        private readonly MethodBase _method;
+        private readonly ContentHeaderBase _header;
+        private readonly ReadOnlyMemory<byte> _body;
+
+        public OutgoingContentCommand(MethodBase method, ContentHeaderBase header, ReadOnlyMemory<byte> body)
         {
-            Method = method;
+            _method = method;
             _header = header;
             _body = body;
         }
 
-        internal void Transmit(ushort channelNumber, Connection connection)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly ReadOnlyMemory<byte> SerializeToFrames(ushort channelNumber, uint frameMax)
         {
-            int maxBodyPayloadBytes = (int)(connection.FrameMax == 0 ? int.MaxValue : connection.FrameMax - EmptyFrameSize);
-            int size = GetMaxSize(maxBodyPayloadBytes);
+            int maxBodyPayloadBytes = (int)(frameMax == 0 ? int.MaxValue : frameMax - Framing.BaseFrameSize);
+            int size = Method.GetRequiredBufferSize() +
+                _header.GetRequiredPayloadBufferSize() +
+                Framing.BodySegment.FrameSize * (maxBodyPayloadBytes == int.MaxValue ? 1 : (_body.Length + maxBodyPayloadBytes - 1) / maxBodyPayloadBytes) + _body.Length
+                + Framing.Method.FrameSize + Framing.Header.FrameSize;
 
             // Will be returned by SocketFrameWriter.WriteLoop
             byte[] rentedArray = ArrayPool<byte>.Shared.Rent(size);
             Span<byte> span = rentedArray.AsSpan(0, size);
```

Comment:
Also here, we can directly pass it into the method (or use rentedArray.AsSpan(offset)) => safes one local variable

Reply:
Not sure I follow

Reply:
Fixed